File: Python/Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        first = ""
        sec = ""
        carry = 0
        result = ListNode()
        head = result
        while (l1 is not None or l2 is not None):
            if (l1 is not None and l2 is not None):
                sum = l1.val + l2.val + carry
            elif (l1 is not None):
                sum = l1.val + carry
            else:
                sum = l2.val + carry
            
            next_val = sum % 10
            carry = 1 if (sum/10 > 0) else 0
            l1 = None if l1 is None else l1.next
            l2 = None if l2 is None else l2.next
            print("sum = %d, next_val = %d, carry = %d" % (sum, next_val, carry))
            result.next = ListNode(next_val)
            result = result.next 
            
        if (carry):
            result.next = ListNode(1)
        
        return head.next
        
        
        

File: Java/Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0; int sum = 0;
        ListNode output = new ListNode(0);
        ListNode node = output;

        while (l1 != null || l2 != null || carry > 0) { 
            int l1Value = l1 != null ? l1.val : 0;
            int l2Value = l2 != null ? l2.val : 0; 
            sum = l1Value + l2Value + carry;
            node.val = sum % 10;

            carry = sum / 10;
            l1 = l1 != null ? l1.next : null; 
            l2 = l2 != null ? l2.next : null; 

            if (l1 != null || l2 != null || carry > 0) {
                node.next = new ListNode(0);
                node = node.next;
            }            
        }
        return output;
    }
}
